Aztec/Classes/Extensions/NSAttributedString+ParagraphRange.swift

```diff
@@ -0,0 +1,184 @@
+import Foundation
+import UIKit
+
+/// Functions for paragraph enumeration and identification.
+///
+extension NSAttributedString {
+       
+    /// Finds the paragraph ranges in the specified string intersecting the specified range.
+    ///
+    /// - Parameters range: The range within the specified string to find paragraphs.
+    ///
+    /// - Returns: An array containing an NSRange for each paragraph intersected by the specified range.
+    ///
+    func paragraphRanges(spanning range: NSRange, includeParagraphSeparator: Bool = true) -> [NSRange] {
+        var paragraphRanges = [NSRange]()
+        let swiftRange = string.range(fromUTF16NSRange: range)
+        let paragraphsRange = string.paragraphRange(for: swiftRange)
+        
+        string.enumerateSubstrings(in: paragraphsRange, options: .byParagraphs) { [unowned self] (substring, substringRange, enclosingRange, stop) in
+            let paragraphRange = includeParagraphSeparator ? enclosingRange : substringRange
+            paragraphRanges.append(self.string.utf16NSRange(from: paragraphRange))
+        }
+        
+        return paragraphRanges
+    }
+    
+    /// Finds the paragraph ranges in the specified string intersecting the specified range.
+    ///
+    /// - Parameters range: The range within the specified string to find paragraphs.
+    ///
+    /// - Returns: An array containing an NSRange for each paragraph intersected by the specified range.
+    ///
+    func paragraphRanges(spanning range: NSRange) -> ([(NSRange, NSRange)]) {
+        var paragraphRanges = [(NSRange, NSRange)]()
+        let swiftRange = string.range(fromUTF16NSRange: range)
+        let paragraphsRange = string.paragraphRange(for: swiftRange)
```

Comment:
I was surprised that 
```swift
string.paragraphRange(for: swiftRange)
```
returns the valid ranges for the entire string, even if the initial range is empty!

I'm thinking that might be worth adding to the inline docs for the parent method? Just reading the unit tests, I'd expect that given a zero point range, we'd get back no matching paragraphs.

Reply:
In terms of the methods we implemented I changed the naming to make them clearer.

The new method name and documentation should make it much clearer what's going on:

```swift
    /// Given a range within the receiver, this method returns an array of ranges for each
    /// paragraph that intercects the provided range.
    ///
    /// - Parameters:
    ///     - range: The initial range that paragraphs must intersect to qualify as valid results.
    ///
    /// - Returns: An array of `ParagraphRange` objects describing the range and the enclosing range
    ///     of each paragraph that intersects the input range.
    ///
    func paragraphRanges(intersecting range: NSRange) -> ([ParagraphRange]) {
```

Regarding the snippet you mention, since that method is not defined / maintained by us, but by Apple I'd say it's up to the reader to know how the method works (as with any other call to standard methods).

Agreed, it's not properly documented in [`StringProtocol`](https://developer.apple.com/documentation/swift/stringprotocol/2923378-paragraphrange), but it works exactly like its [NSString](https://developer.apple.com/documentation/foundation/nsstring/1408548-paragraphrange) counterpart.